lib/inv_gcr_quda.cpp

```diff
@@ -264,8 +284,9 @@ namespace quda {
     const int maxResIncrease = param.max_res_increase; // check if we reached the limit of our tolerance
     const int maxResIncreaseTotal = param.max_res_increase_total;
 
-    double heavy_quark_res = 0.0; // heavy quark residual
-    if(use_heavy_quark_res) heavy_quark_res = sqrt(blas::HeavyQuarkResidualNorm(x,r).z);
+    std::vector<double> heavy_quark_res(b.size()); // heavy quark residual
+    if (use_heavy_quark_res)
+      for (auto i = 0u; i < b.size(); i++) heavy_quark_res[i] = sqrt(blas::HeavyQuarkResidualNorm(x, r)[i].z);
```

Comment:
Not that I'm stressed about how efficiently we compute the HQRes, but this recomputes the HQR multiple times, I think this should be more like:

```
std::vector<double> heavy_quark_res(b.size()); // heavy quark residual
if (use_heavy_quark_res) {
  auto hqr = blas::HeavyQuarkResidualNorm(x, r);
  for (auto i = 0u; i < b.size(); i++) heavy_quark_res[i] = sqrt(hqr[i].z);
}
```

Reply:
f0d9f3c8dc79bdb4ed751b34193262c517a6b287